coresimd/x86_64/mod.rs

```diff
@@ -38,3 +43,64 @@ pub use self::bswap::*;
 
 mod rdrand;
 pub use self::rdrand::*;
+
+/// Compare and exchange 16 bytes (128 bits) of data atomically.
+///
+/// This intrinsic corresponds to the `cmpxchg16b` instruction on x86_64
+/// processors. It performs an atomic compare-and-swap, updating the `ptr`
+/// memory location to `val` if the current value in memory equals `old`.
+///
+/// # Return value
+///
+/// This function returns the previous value at the memory location. If it is
+/// equal to `old` then the memory was updated to `new`.
+///
+/// # Memory Orderings
+///
+/// This atomic operations has the same semantics of memory orderings as
+/// `AtomicUsize::compare_exchange` does, only operating on 16 bytes of memory
+/// instead of just a pointer.
+///
+/// For more information on memory orderings here see the `compare_exchange`
+/// documentation for other `Atomic*` types in the standard library.
+///
+/// # Unsafety
+///
+/// This method is unsafe because it takes a raw pointer and will attempt to
+/// read and possibly write the memory at the pointer. The pointer must also be
+/// aligned on a 16-byte boundary.
+///
+/// This method also requires the `cmpxchg16b` CPU feature to be available at
+/// runtime to work correctly, and executing this function if the CPU doesn't
```

Comment:
I would word this as "...correctly. If the CPU running the binary does not actually support `cmpxchg16b` and the program enters an execution path that eventually would reach this function the behavior is undefined." 

As in, the problem here is that an optimizing compiler is able to assume that the CPU supports `cmpxchg16b` in all execution paths that unconditionally reach this function, and is able to emit machine code under that assumption in those. That is, once the program enters such an execution path, all bets are off: there are no guarantees that this function will ever be reached.

cc @RalfJ